modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/passthru/ServerWorker.java

```diff
@@ -189,10 +190,19 @@ public SOAPEnvelope handleRESTUrlPost(String contentTypeHdr) throws FactoryConfi
         // When POST request doesn't contain a Content-Type,
         // recipient should consider it as application/octet-stream (rfc2616)
         if (contentType == null || contentType.isEmpty()) {
-            contentType = PassThroughConstants.APPLICATION_OCTET_STREAM;
+            String httpMethod = (String) msgContext.getProperty(HTTP_METHOD);
             // Temp fix for https://github.com/wso2/product-ei/issues/2001
-            if (HTTPConstants.HTTP_METHOD_GET.equals(msgContext.getProperty(HTTP_METHOD))) {
+            // Else if block is added to prevent the content-type assignment as application/octet-stream
+            // for all incoming requests. Current fix sets content-type as application/octet-stream only
+            // for POST requests which has body and no content-type
+            // as per spec: https://tools.ietf.org/html/rfc2616#page-43
+            if (HTTPConstants.HTTP_METHOD_GET.equals(httpMethod)) {
                 contentType = HTTPConstants.MEDIA_TYPE_X_WWW_FORM;
+            } else if (HTTPConstants.HTTP_METHOD_POST.equals(httpMethod)) {
+                if (request.isEntityEnclosing() &&
+                        ((HttpEntityEnclosingRequest) request.getRequest()).getEntity().getContentLength() != 0) {
```

Comment:
What would happen getEntity returns null?

Reply:
isEntityEnclosing() returns whether the request has body or not - [1]. When getEntity() returns null, request does not have body, hence if condition won't get executed
[1] https://github.com/wso2/wso2-synapse/blob/master/modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/passthru/SourceRequest.java#L49